[1단계 - 장기] 리원(김지윤) 미션 제출합니다.

awrion3 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4122)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

1. 초기 설계 방향과 다르게 실제 애플리케이션을 구현하게 되어 객체 지향 프로그래밍보다는 최소한의 요구사항(장기 보드 초기화, 장기 기물 이동)만을 준수하고자 하였습니다.
2. 기존 구현 코드를 제거하고 처음부터 새롭게 구현하는 과정을 단시간에 수행하게 되어, 두번째 구현하는 과정에서는 TDD를 적용하지 못하였습니다.
3. 최소한의 기능만 구현한 만큼 향후 장기 기물의 움직임과 관련한 검증 로직, 점수 계산과 반복 실행 등 전체적인 장기 게임 로직 등을 추가할 예정입니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

- 초기 설계와 달라진 구현
초기에는 실제 게임을 염두에 두고 턴당 사용자로부터 말 이름을 입력 받으면, 이동 가능한 위치를 보여준 후 실제로 이동할 좌표를 선택하면 반영되게 구현하고자 했습니다.
다만 관련 로직을 구현하는 과정에서 지나치게 로직이 복잡해져 기존 설계와 구현 내역을 취소하게 되었습니다. 
그리고 이번에는 사용자가 실제로 이동할 좌표를 입력하게 한 후 이를 검증하고 반영하는 방식으로 변경하게 되었습니다.

```
안녕하세요, 이프!
이번 미션의 경우 코드 구현 과정에서 초기 설계의 경우 경로 탐색 로직을 단순화 하는 과정에서 다소 어려움을 겪어, 
기존 구현을 폐기하고 단시간에 새롭게 변경하게 되어 부족한 부분이 많은 점 양해 부탁드립니다,
향후 테스트 코드와 비즈니스 로직을 더 추가해 나가보도록 하겠습니다!
시간 내어 리뷰 해주셔서 감사합니다 ☺️
```

kwonkeonhyeong committed:
docs: README.md 기능 구현 목록 수정

kwonkeonhyeong committed:
test: position 객체 생성 및 갱신 테스트 추가

kwonkeonhyeong committed:
feat: PositionX, PositionY 생성 및 검증 구현

kwonkeonhyeong committed:
feat: Position 객체 생성 및 갱신 기능 구현

kwonkeonhyeong committed:
feat: 기물 이름에 맞는 Movement 반환 기능 구현

kwonkeonhyeong committed:
feat: 말의 이동 가능 position 목록 생성 기능 구현

kwonkeonhyeong committed:
test: 진영에 따른 병졸 이동 가능 목록 반환 확인 테스트 추가

kwonkeonhyeong committed:
refactor: 진영에 따른 병 졸 이동 가능 목록 반환 기능 개선

kwonkeonhyeong committed:
feat: 차의 이동 가능 목록 반환 기능 구현

kwonkeonhyeong committed:
feat: 각 말의 이동 가능 위치 목록 반환 및 기본 위치 생성 기능 구현

kwonkeonhyeong committed:
feat: 선택한 말의 실제 시동 가능 위치 확인 메서드 개요 작성

kwonkeonhyeong committed:
Changes by kyun

kwonkeonhyeong committed:
feat: 상의 장애물에 따른 이동 위치 계산 구현

kwonkeonhyeong committed:
Changes by kyun

kwonkeonhyeong committed:
refactor: 프로그램 구조 전면 개편

kwonkeonhyeong committed:
feat: Position X,Y 좌표 및 Position 생성 로직 구현

kwonkeonhyeong committed:
feat: 장기판 초기화 기능 구현

kwonkeonhyeong committed:
feat: 도착 지점에 말을 이동시킬 수 있는지 검증 하는 기능 구현

kwonkeonhyeong committed:
feat: 말 이동 기능 구현

kwonkeonhyeong committed:
feat: 말 이동 경로 검증 기능 구현 (포 제외)

kwonkeonhyeong committed:
feat: 포 이동 여부 검증 구현

kwonkeonhyeong committed:
docs: README.md 기능 구현 목록 수정 및 체크

sinb57 commented on src/main/java/janggi/position/PositionX.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Objects;
+
+public class PositionX {
```
`PositionX`랑 `PositionY`는 만든 목적이 무엇인가요?

awrion3 replied:
음, 다시 생각해보니 각 x와 y 좌표에 대한 범위 검증 등은 현재 다른 곳에서 진행 중인 만큼, Position(x, y)에서 각각 int 값을 저장하는 형태로도 충분히 의미를 전달할 수 있다고 생각된다고 느껴지네요. 굳이 PositionX, PositionY를 추가적으로 사용하는 대신, int x, y를 사용하는 방향으로 리팩토링 했습니다!

sinb57 commented on src/main/java/janggi/team/Team.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public static Team from(String teamName) {
+        return Arrays.stream(Team.values()).filter(team -> team.getValue().equals(teamName))
```
getter를 사용하지 말고, 메시지를 던져보세요~

awrion3 replied:
아 네, getter 대신 match 여부를 확인하는 메서드를 추가하는 방식으로 리팩토링 해보겠습니다!

sinb57 commented on src/main/java/janggi/game/GameManager.java:
```diff
@@ -0,0 +1,101 @@
+        }
+        if (pieceName.equals("P")) {
+            boolean isLegal = false;
+            if (teamName.equals(Team.CHO)) {
+                isLegal = boardCho.isLegalMoveForCannon(positionsOnPath);
+            }
+            if (teamName.equals(Team.HAN)) {
+                isLegal = boardHan.isLegalMoveForCannon(positionsOnPath);
+            }
```
`if (teamName.equals(Team.CHO)) {`, `if (teamName.equals(Team.HAN)) {` 분기로직이 반복되네요! 다형성을 활용해보는 건 어떨까요

awrion3 replied:
네, 되도록 유사한 분기 로직이 반복되는 부분이 없게끔 리팩토링 해보겠습니다!

sinb57 commented on README.md:
```diff
@@ -1,3 +1,53 @@
+  - [ ] 생사 여부 확인
+  - [ ] 이동 규칙 저장
+- [4] 위치
+  - 좌표 값(x, y) 저장
```
실행해보니 어떤 입력을 해야할지 막막했어요. 어디가 0,0 인지 기준 잡기 어렵더라고요. 장기 판의 x축과 y축에 좌표를 보여주면 어떨까요? 아래 사진 속 체스판 처럼요.

<img src="https://github.com/user-attachments/assets/8399ea1a-42ed-42fa-afb8-09f7498ee5ca" width=400>


그리고 README 가 사용 설명서의 역할을 수행할 수 있도록, 내용을 보충해주시면 좋겠어요~ 입력 예시를 추가해보면 어떨까요?

awrion3 replied:
오 네, 체스판과 같이 각 축의 좌표가 보이게끔 구조를 개선 해볼게요! 그리고 readme에 입력 예시를 작성해 볼 생각은 못했네요, 추가해보겠습니다! 그리고 input을 요청할 때도 입력 예시를 함께 제시하는 방향으로 수정해보도록 하겠습니다!

sinb57 commented on src/main/java/janggi/board/BoardOption.java:
```diff
@@ -0,0 +1,43 @@
+
+    public static BoardOption of(Team team, String option) {
+        return Arrays.stream(BoardOption.values())
+                .filter(value -> value.team == team && value.option.equals(option))
```
메서드를 분리해보면 어떨까요

awrion3 replied:
오 네, 그럼 해당 부분에서 현재와 같이 메서드 체이닝을 통해 구현하는 방식 외에도 예외를 던지는 부분을 추가적으로 분리하는 방식으로 리팩토링 하는 방법을 말씀하신 것인가요?

sinb57 replied:
orElseThrow 까지 메서드 분리할 필요는 없고요. `value.teamName == teamName && value.option.equalsIgnoreCase(option)`를 메서드 분리해서 의도가 읽히는 코드를 작성하길 바랬어요

sinb57 commented on src/main/java/janggi/board/BoardNavigator.java:
```diff
@@ -0,0 +1,30 @@
+        int endX = end.getX();
+        int endY = end.getY();
+
+        int dx = endX - startX;
+        int dy = endY - startY;
```
약어 사용은 지양해주세요~

awrion3 replied:
네, 약어 사용한 해당 부분 수정했습니다!

sinb57 commented on src/main/java/janggi/game/GameManager.java:
```diff
@@ -0,0 +1,101 @@
+        this.output = new Output();
+    }
+
+    public void run() {
```
오.. 리팩터링 해주실 거죠?ㅎ

awrion3 replied:
아 네, 메서드를 좀 더 세분화해서 리팩토링 해보겠습니다!

sinb57 commented on src/main/java/janggi/game/GameManager.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    public Position parsePosition(String piecePosition) {
+        Pattern pattern = Pattern.compile("\\d+"); // \\d+는 연속된 숫자를 추출
```
이펙티브자바 (p32. 아이템 6)를 읽어보시면 좋을 듯 해요~

awrion3 replied:
네, 참고할 만한 내용 감사합니다. Pattern 대신 Stream을 사용해 변환하는 방향으로 수정했습니다!

sinb57 reviewed (changes requested):
안녕하세요 리원! 만나서 반가워요ㅎ Level1 마지막 미션의 리뷰어를 맡은 이프 예요!
장기 미션 하시느라 고생 많으셨어요. 첫번째 리뷰 요청 기한이 빠듯했죠?ㅠ
코드 살펴봤는데, 구현 및 설계가 완전치 않아서 리뷰 드릴 내용이 많지 않네요..
마저 완성하신 후 요청주시면 다시 봐드릴게요!

awrion3 committed:
refactor: remove old class files

awrion3 committed:
refactor: relocate and add methods for board

- validate range, turn, find positions for Board, convert setup for BoardSetup, update position for Position

awrion3 committed:
feat: add palace area and check position method

awrion3 committed:
fear: define piece movement validation, add palace movement

awrion3 committed:
feat: apply abstract on team class, add validate palace movement, and update caught status

- add calculate team score, add convert team name

awrion3 committed:
refactor: use void instead of returning boolean, apply constants

- use team instead of team name for validating team turn in Board

awrion3 committed:
feat: add validate and repeat input, split methods in manager class

- apply coordinate format in output

awrion3 committed:
fix: add trim in parse position method

- change variable name in input

awrion3 committed:
docs: update feature list in README.md

- add input samples

awrion3 committed:
refactor: relocate update status method to piece class

awrion3 committed:
refactor: relocate check inside palace method to palace class

- use ignored value as variable name in palace area enum

sinb57 commented on src/main/java/janggi/board/Position.java:
```diff
@@ -0,0 +1,8 @@
+
+public record Position(int x, int y) {
+
+    public Position update(Position position) {
+        return new Position(position.x(), position.y());
```
무슨 목적의 메서드인지 잘 모르겠어요~ Position이 record로 선언되었기에 불변인데요, position 객체를 인자로 받아 새로 생성하는 copy 작업이 왜 필요한지 모르겠어요

awrion3 replied:
앗 네, 해당 메서드는 굳이 필요 없겠네요, 삭제하는 방향으로 리팩토링 했습니다!

sinb57 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public void validatePieceRange(Position destination) {
+        boolean isValidRange = destination.x() >= BOARD_AXIS_MIN && destination.x() <= BOARD_X_AXIS_MAX
+                && destination.y() >= BOARD_AXIS_MIN && destination.y() <= BOARD_Y_AXIS_MAX;
+        if (!isValidRange) {
+            throw new IllegalArgumentException(INVALID_RANGE);
+        }
+    }
```
Position 객체 생성 시 검증하도록, Position 클래스가 지니면 좋을 듯 한데 어떤가요?

awrion3 replied:
오 네, 그럴 수도 있겠네요. Position을 생성할 때 x, y 좌표의 값이 검증되도록 변경했습니다!

sinb57 commented on src/main/java/janggi/team/TeamCho.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TeamCho extends Team {
+
```
Init 과정 때문에 상속 구조를 취한 것 같은데요. 상속 구조 보다는, 팀을 생성하는 팩토리 클래스를 만들어두는 것은 어떨까요? 현재는 상속으로 얻는 이점이 무엇인지 모르겠어요~

awrion3 replied:
오 네, 상속 구조보다는 팩토리 클래스를 적용해보는 것이 더 낫겠군요, 참고해서 수정해보겠습니다!

sinb57 commented on src/main/java/janggi/view/Input.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Scanner;
+import java.util.function.Supplier;
+
+public class Input {
```
view 패키지 안에 두 클래스가 있길래, 입력은 Input에서 이뤄지고 출력은 Output에서 이뤄진다고 생각했어요. 그런데 Input에서 입력을 위한 안내 문구 출력이 이뤄지네요. Input와 Output의 차이는 무엇인가요? View를 Input / Output으로 구분지을 필요가 있나요?

awrion3 replied:
음 네, 말씀하신대로 입력을 위한 안내 문구 출력이 존재하고 있어서 view를 굳이 구분할 필요는 없다고 생각해요. 다만 하나의 view로 합치게 되는 경우, 입출력 코드의 구분 및 향후 관리가 명확해지지 않을 수도 있다고 생각해서 현재와 같은 구조를 유지해보게 되었습니다!

sinb57 replied:
프로젝트가 커지고 입출력이 많아지면, Input View와 Output View가 비대해지고 관리가 어려워질텐데요. 입력과 출력이 아닌, 화면을 기준으로 View를 분리하면 어떨까요? 키오스크가 여러 화면을 가진 것 처럼요!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -0,0 +1,134 @@
+
+    public void run() {
+        TeamCho teamCho = repeatInput(
+                () -> new TeamCho(BoardSetup.of(TeamName.CHO, input.readPositionOption(TeamName.CHO))));
```
TeamCho를 생성하면서 TeamName.CHO를 인자로 넣는 행위가 이상해보여요~ 그렇다면 TeamName.HAN을 가지고서 TeamCho를 생성할 수 있다는 의미인데, 의도된 내용인지 궁금해요

awrion3 replied:
음, 그렇게 될 수도 있겠네요. 해당 부분은 input 값만을 사용해서 보드가 세팅될 수 있도록 변경해보겠습니다!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -0,0 +1,134 @@
+    private Team validateTurn(Board board, TeamCho teamCho, TeamHan teamHan, Team oldTeam) {
+        while (true) {
+            try {
+                TeamName teamName = repeatInput(() -> TeamName.from(input.readPieceTeamName()));
```
팀을 선택하여 기물을 이동하는 것이 아니라, 번갈아가며 기물을 이동하는 것이 장기의 규칙 아닌가 싶어요~

awrion3 replied:
아 네, 팀 선택을 통한 방식이 아닌, 좀 더 장기 게임처럼 번갈아서 기물 이동을 하게끔 수정해보겠습니다!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    private Position parsePosition(String coordinates) {
+        List<Integer> pieceCoordinates = Arrays.stream(coordinates.replaceAll(REGEX_PATTERN, "").split(DELIMITER_COMMA))
```
이 행위는 입출력 요구사항에 의존할텐데요. view 패키지 내 클래스가 아닌, Manager 도메인 객체가 담당하는 것이 옳을까요?

awrion3 replied:
아, view 패키지 내 클래스에 있는 쪽이 더 낫다고 생각되네요, 해당 위치로 옮기도록 해보겠습니다!

sinb57 commented on src/main/java/janggi/piece/Elephant.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void validateMovement(Position currentPosition, Position destination, PalaceArea palaceArea) {
+        int offsetX = Math.abs(currentPosition.x() - destination.x());
+        int offsetY = Math.abs(currentPosition.y() - destination.y());
+
+        boolean isValidMove = offsetX == OFFSET_TWO && offsetY == OFFSET_THREE
+                || offsetX == OFFSET_THREE && offsetY == OFFSET_TWO;
```
값을 가져와서 연산하는 것이 아니라, Position에 메시지를 던지는 방식은 어떤가요? currentPosition에게 destination과 OFFSET_TWO를 같이 전달하여 Boolean을 반환 받는 거죠

awrion3 replied:
음, 말씀해주신 방식으로 하면 Position이 계산을 담당하게 되어 좀 더 깔끔한 구조가 될 수 있을 것 같아요. 다만 Position은 좌표를 저장하거나 해당 좌표의 범위를 검증하는 역할만을 부여하고, 각 기물의 이동 규칙은 Piece가 정의할 수 있도록 구성해보고 싶어서 현재와 같은 방법을 선택해보게 되었어요. 또한 offset 계산의 경우에는  Position이 담당하도록 변경해보게 되었습니다.

sinb57 commented on src/main/java/janggi/palace/Palace.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class Palace {
```
> 궁(宮) 또는 궁성(宮城)의 영역은 구현하지 마라.

1단계 요구사항이 아닌데 2단계 미리 반영하신걸까요?

awrion3 replied:
아 네, 코드 구현 과정에서 2.1 단계까지 미리 반영해서 작성하게 된 것 같습니다!

sinb57 replied:
그렇군요! 앞으로는 요구사항을 미리 반영하지 않으셨으면 좋겠어요. 단계가 나눠져 있는 이유가 있고 이를 통해 학습을 유도하는 바가 있는데, 미리 앞서서 하게 되면 기능 추가에 따른 구조 변경이나 리팩터링 등의 기회를 놓치게 돼요.

sinb57 reviewed (changes requested):
안녕하세요 리원~ 고생 많으셨어요! 추가 코멘트 남겼으니 확인 부탁드리고,
`객체지향 생활 체조 원칙` 규칙 맞춰서 리팩터링 부탁드려요~

awrion3 committed:
refactor: apply factory instead of abstract class for team

- remove update method, add range validation for position

awrion3 committed:
refactor: change board setup parameters, relocate parse position to input class

- adjust code to reflect changes for board, output, piece class

awrion3 committed:
refactor: remove team turn request and validation

awrion3 committed:
refactor: move calculate offset to position

- add match team name method

sinb57 commented on src/main/java/janggi/board/Position.java:
```diff
@@ -0,0 +1,27 @@
+        validateRange(x, y);
+    }
+
+    private void validateRange(int x, int y) {
```
인자를 받지 않아도 괜찮을 텐데요 어떻게 생각하시나요?

awrion3 replied:
아 네, 레코드 생성자 내에서 사용되는 만큼 굳이 인자로 받지 않아도 되었군요. 그리고 입력 패턴에서 x, y 값의 범위를 이미 검증하고 있어서, 동일한 검증을 Position에서 다시 진행하지 않아도 된다고 판단해 제거하는 방향으로 수정했습니다!

sinb57 reviewed (approved):
안녕하세요 리원~ 2단계 진행을 위해 머지할게요!
추가 코멘트가 몇 없지만, 2단계에서 반영해주셨으면 하는 부분이 있어서 따로 적어보아요.

- 객체지향 생활체조 요구사항을 충족해주세요. 특히 Manager 클래스는 꼭 충족해주세요.
- Palace 추상 클래스를 두어 상속 구조를 취한 이유가 궁금해요. 얻고자 한 이점이 무엇인가요
- view를 제외한 모든 클래스들의 public 메서드는 테스트 코드를 작성해주세요. 특히 기물에 대해서요
- Piece 들의 validateMovement 로직이 잘 읽히는 구조가 아니에요. 가독성을 높일 수 없는지 고민해보셔요
